### PR TITLE
feature: convert indexing service to kotlin

### PR DESCRIPTION
**Pull request by hsiliev**
Convert to Kotlin and adapt tests

Solves: PZ-3714

### CODE REVIEW

File: src/main/java/net/atos/zac/app/planitems/PlanItemsRESTService.java [hunk 1]
```diff
@@ -64,7 +64,7 @@
 import net.atos.zac.shared.helper.OpschortenZaakHelper;
 import net.atos.zac.util.DateTimeConverterUtil;
 import net.atos.zac.util.UriUtil;
-import net.atos.zac.zoeken.IndexeerService;
+import net.atos.zac.zoeken.IndexingService;
```
Comment (edgarvonk) [inline src/main/java/net/atos/zac/app/planitems/PlanItemsRESTService.java:67]: While you're at it, could also consider renaming the `zoeken` package to `search`?

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+import org.flowable.task.api.Task
+import java.io.IOException
+import java.util.UUID
+import java.util.logging.Logger
+
+@Singleton
+@Transactional
+@AllOpen
+@Suppress("TooManyFunctions")
+class IndexingService @Inject internal constructor(
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:45]: Why `internal`?

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+    @param:Any private val converterInstances: Instance<AbstractZoekObjectConverter<out ZoekObject>>,
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:46]: This looks strange? Is this `@param:Any` needed?

**Review by edgarvonk — COMMENTED** [review]
just some quick remarks. I know it's just a draft PR. ;-)

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt (hunk 2 above)
Comment (hsiliev) [reply src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:45]: Converter added this. Thanks for noticing! Removed

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt (hunk 3 above)
Comment (hsiliev) [reply src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:46]: Not needed. It's the default. Converter issues I think

File: src/main/java/net/atos/zac/app/planitems/PlanItemsRESTService.java (hunk 1 above)
Comment (hsiliev) [reply src/main/java/net/atos/zac/app/planitems/PlanItemsRESTService.java:67]: Not everything is converted, so better keep it still

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+    private val drcClientService: DrcClientService,
+    private val flowableTaskService: FlowableTaskService
+) {
+    companion object {
+        const val SOLR_CORE: String = "zac"
+
+        private const val SOLR_MAX_RESULT = 100
+        private const val TAKEN_MAX_RESULTS = 50
+
+        private val LOG: Logger = Logger.getLogger(IndexingService::class.java.name)
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:56]:
```suggestion
        private val LOG = Logger.getLogger(IndexingService::class.java.name)
```

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+@AllOpen
+@Suppress("TooManyFunctions")
+class IndexingService @Inject constructor(
+    private val converterInstances: Instance<AbstractZoekObjectConverter<out ZoekObject>>,
+    private val zrcClientService: ZrcClientService,
+    private val drcClientService: DrcClientService,
+    private val flowableTaskService: FlowableTaskService
+) {
+    companion object {
+        const val SOLR_CORE: String = "zac"
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:51]:
```suggestion
        const val SOLR_CORE = "zac"
```

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+import org.apache.solr.common.SolrDocument
+import org.apache.solr.common.params.CursorMarkParams
+import org.eclipse.microprofile.config.ConfigProvider
+import org.flowable.task.api.Task
+import java.io.IOException
+import java.util.UUID
+import java.util.logging.Logger
+
+@Singleton
+@Transactional
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:41]: Really transactional? Only needed if we are updating our own database I would think and not for updating a Solr index?

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+            if (cursorMark == nextCursorMark) {
+                done = true
+            } else {
+                cursorMark = nextCursorMark
+            }
+        }
+    }
+
+    private fun reindexAllZaken() {
+        val listParameters = ZaakListParameters()
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:244]: .apply { .. } ?

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+        val listParameters = EnkelvoudigInformatieobjectListParameters()
+        listParameters.page = ZGWApiService.FIRST_PAGE_NUMBER_ZGW_APIS
+        var hasMore: Boolean
+        do {
+            hasMore = reindexInformatieobjecten(listParameters)
+            listParameters.page += 1
+        } while (hasMore)
+    }
+
+    private fun reindexAllTaken() {
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:264]: Looks a bit clumsy? Maybe Copilot has a nice suggestion to improve this code? ;-)

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+        listParameters: EnkelvoudigInformatieobjectListParameters
+    ): Boolean {
+        val results = drcClientService.listEnkelvoudigInformatieObjecten(listParameters)
+        indexeerDirect(
+            results.results
+                .map {
+                        enkelvoudigInformatieObject ->
+                    URIUtil.parseUUIDFromResourceURI(enkelvoudigInformatieObject.url)
+                }
+                .map { obj: UUID -> obj.toString() },
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:302]:
```suggestion
                .map { it.toString() },
```

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+
+    private fun reindexTaken(page: Int, numberOfTasks: Long): Boolean {
+        val firstResult = page * TAKEN_MAX_RESULTS
+        val tasks = flowableTaskService.listOpenTasks(
+            TaakSortering.CREATIEDATUM,
+            SorteerRichting.DESCENDING,
+            firstResult,
+            TAKEN_MAX_RESULTS
+        )
+        indexeerDirect(tasks.map { obj: Task -> obj.id }, ZoekObjectType.TAAK, false)
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:323]:
```suggestion
        indexeerDirect(tasks.map { i.id }, ZoekObjectType.TAAK, false)
```

and other places

**Review by edgarvonk — CHANGES_REQUESTED** [review]

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+
+        private const val SOLR_MAX_RESULT = 100
+        private const val TAKEN_MAX_RESULTS = 50
+
+        private val LOG = Logger.getLogger(IndexingService::class.java.name)
+
+        private val reindexingViewfinder: MutableSet<ZoekObjectType> = HashSet()
+    }
+
+    private val solrClient: SolrClient = Http2SolrClient.Builder(
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:60]: If you want to mimic how it was done in the Java code (= constructor time) then this should be in an `init` block. But not sure if this is really needed in this case. 
https://blog.mindorks.com/understanding-init-block-in-kotlin/

Comment (hsiliev) [reply src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:60]: Seems to work without it, so decided not to use `init` here

File: src/main/kotlin/net/atos/zac/app/search/rest/IndexingRestService.kt [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+@NoArgConstructor
+@AllOpen
+class IndexingRestService @Inject constructor(
+    private val indexingService: IndexingService
+) {
+    @GET
+    @Path("herindexeren/{type}")
+    fun reindex(@PathParam("type") type: ZoekObjectType) {
+        indexingService.reindex(type)
+    }
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/app/search/rest/IndexingRestService.kt:34]:
```suggestion
    fun reindex(@PathParam("type") type: ZoekObjectType) = indexingService.reindex(type)
```

also in other places. just makes the code a bit easier to read

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+) {
+    companion object {
+        const val SOLR_CORE = "zac"
+
+        private const val SOLR_MAX_RESULT = 100
+        private const val TAKEN_MAX_RESULTS = 50
+
+        private val LOG = Logger.getLogger(IndexingService::class.java.name)
+
+        private val reindexingViewfinder: MutableSet<ZoekObjectType> = HashSet()
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:54]:
```suggestion
        private val reindexingViewfinder = mutableSetOf<ZoekObjectType>()
```

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+    }
+
+    private val solrClient: SolrClient = Http2SolrClient.Builder(
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:57]: In an `init {}` block?

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+     * Add a list of objectIds to the Solr index and optionally performs a (hard) Solr commit so
+     * that the Solr index is updated immediately.
+     * Beware that hard Solr commits are relatively expensive operations.
+     *
+     * @param objectIds     the list of object ids to be indexed
+     * @param objectType    the object type
+     * @param performCommit whether to perform a hard Solr commit
+     */
+    fun indexeerDirect(objectIds: List<String>, objectType: ZoekObjectType, performCommit: Boolean) =
+        addToSolrIndex(objectIds.map { objectId: String -> getConverter(objectType).convert(objectId) }, performCommit)
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:83]:
```suggestion
        addToSolrIndex(objectIds.map { getConverter(it).convert(it) }, performCommit)
```

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt [hunk 16]
```diff
@@ -0,0 +1,319 @@
...
+        addOrUpdateInformatieobject(
+            UriUtil.uuidFromURI(
+                zrcClientService.readZaakinformatieobject(zaakinformatieobjectUUID).informatieobject
+            )
+        )
+    }
+
+    fun addOrUpdateTaak(taskID: String) {
+        indexeerDirect(taskID, ZoekObjectType.TAAK, false)
+    }
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:128]:
```suggestion
    fun addOrUpdateTaak(taskID: String) = indexeerDirect(taskID, ZoekObjectType.TAAK, false)
```

etc

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt [hunk 17]
```diff
@@ -0,0 +1,303 @@
...
+        return false
+    }
+
+    private fun logTypeMessage(objectType: ZoekObjectType, message: String) =
+        LOG.info("[$objectType] $message")
+
+    private fun logProgress(objectType: ZoekObjectType, voortgang: Long, grootte: Long) =
+        logTypeMessage(objectType, "reindexed: $voortgang / $grootte")
+
+    private fun <T> wrapInIndexingException(fn: () -> T): T {
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:294]: Bit of a weird construct this I find but let's leave this for now I guess.

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt [hunk 18]
```diff
@@ -0,0 +1,303 @@
...
+    private val flowableTaskService: FlowableTaskService
+) {
+    companion object {
+        const val SOLR_CORE = "zac"
+
+        private const val SOLR_MAX_RESULT = 100
+        private const val TAKEN_MAX_RESULTS = 50
+
+        private val LOG = Logger.getLogger(IndexingService::class.java.name)
+        private val reindexingViewfinder = mutableSetOf<ZoekObjectType>()
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:52]: so I think a plain hashmap is not thread safe. Better to use a Java ConcurrentHashMap here I think or maybe an Kotlin equivalent (something Atomic). Maybe we will never run into such concurrency issues but I think better be safe here.

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt [hunk 19]
```diff
@@ -0,0 +1,303 @@
...
+class IndexingService @Inject constructor(
+    private val converterInstances: Instance<AbstractZoekObjectConverter<out ZoekObject>>,
+    private val zrcClientService: ZrcClientService,
+    private val drcClientService: DrcClientService,
+    private val flowableTaskService: FlowableTaskService
+) {
+    companion object {
+        const val SOLR_CORE = "zac"
+
+        private const val SOLR_MAX_RESULT = 100
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:48]:
```suggestion
        private const val SOLR_MAX_RESULTS = 100
```

?

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt [hunk 20]
```diff
@@ -0,0 +1,303 @@
...
+            reindexingViewfinder.remove(objectType)
+        }
+    }
+
+    fun addOrUpdateZaak(zaakUUID: UUID?, inclusiefTaken: Boolean) {
+        indexeerDirect(zaakUUID.toString(), ZoekObjectType.ZAAK, false)
+        if (inclusiefTaken) {
+            flowableTaskService.listOpenTasksForZaak(zaakUUID)
+                .map { it.id }
+                .forEach { id -> this.addOrUpdateTaak(id) }
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:112]:
```suggestion
                .forEach(this::addOrUpdateTaak)
```

File: src/main/kotlin/net/atos/zac/app/search/rest/IndexingRestService.kt (hunk 12 above)
Comment (bas-info-nl) [reply src/main/kotlin/net/atos/zac/app/search/rest/IndexingRestService.kt:34]: I can't see this suggestion in the code?

File: src/main/kotlin/net/atos/zac/app/search/rest/IndexingRestService.kt [hunk 21]
```diff
@@ -0,0 +1,41 @@
+/*
+ * SPDX-FileCopyrightText: 2022 Atos
```
Comment (bas-info-nl) [inline src/main/kotlin/net/atos/zac/app/search/rest/IndexingRestService.kt:2]:
```suggestion
 * SPDX-FileCopyrightText: 2022 Atos, 2024 Lifely
```

File: src/main/kotlin/net/atos/zac/app/search/rest/IndexingRestService.kt [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+    private val indexingService: IndexingService
+) {
+    @GET
+    @Path("herindexeren/{type}")
+    fun reindex(@PathParam("type") type: ZoekObjectType) {
+        indexingService.reindex(type)
+    }
+
+    @POST
+    @Path("commit-pending-changes-to-search-index")
```
Comment (bas-info-nl) [inline src/main/kotlin/net/atos/zac/app/search/rest/IndexingRestService.kt:37]: I've seen different ways that we write these paths, snake-case like this and camel-case in other places. Is there a standard we want to use (and enforce) ?

File: src/main/kotlin/net/atos/zac/zoeken/IndexingException.kt [hunk 23]
```diff
@@ -0,0 +1,6 @@
+package net.atos.zac.zoeken
```
Comment (bas-info-nl) [inline src/main/kotlin/net/atos/zac/zoeken/IndexingException.kt:1]:
```suggestion
/*
 * SPDX-FileCopyrightText: 2024 Lifely
 * SPDX-License-Identifier: EUPL-1.2+
 */ 
package net.atos.zac.zoeken
```

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt [hunk 24]
```diff
@@ -0,0 +1,303 @@
+/*
+ * SPDX-FileCopyrightText: 2022 Atos
```
Comment (bas-info-nl) [inline src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:2]:
```suggestion
 * SPDX-FileCopyrightText: 2022 Atos, 2024 Lifely
```

**Review by bas-info-nl — APPROVED** [review]
Just some minor comments and suggestions.

File: src/main/kotlin/net/atos/zac/app/search/rest/IndexingRestService.kt (hunk 12 above)
Comment (hsiliev) [reply src/main/kotlin/net/atos/zac/app/search/rest/IndexingRestService.kt:34]: Fixed now

File: src/main/kotlin/net/atos/zac/app/search/rest/IndexingRestService.kt (hunk 22 above)
Comment (hsiliev) [reply src/main/kotlin/net/atos/zac/app/search/rest/IndexingRestService.kt:37]: We do have such convention:
https://github.com/infonl/dimpact-zaakafhandelcomponent/blob/main/CONTRIBUTING.md#coding-conventions

File: src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt (hunk 17 above)
Comment (hsiliev) [reply src/main/kotlin/net/atos/zac/zoeken/IndexingService.kt:294]: Tried to get rid of code duplication, but indeed this is not quite standard

**Review by edgarvonk — APPROVED** [review]
nice!

**Pull request merged**